Created Button to mark a Package As Received

mbetances805 opened the pull request:
Creating this PR to review work in progress.

I created a button to mark a package from Packed to Received. I could not figure out how to enable just one column in the table to be editable with react-bootstrap-table. Including 'cellEdit' would make the entire table editable.

Question: Do the individual items in the package also need to be updated to 'Received? At this time only the package is updated to 'Received'.

I had an issue committing my code due to the 'admins can set other users as admins' test. That test kept failing, so I commented it out.

Let me know if this is in the right direction.

Thanks!

jspaine commented on client/modules/food/components/packing/Packages.js:
```diff
@@ -39,14 +53,37 @@ class Packages extends Component {
+    )
+  }
+    
+  getPackageButton = (_, foodPackage) => {
```
Why not do this in `getActionButtons` above? Then they can be in the same table column

jspaine commented on client/modules/food/reducers/packing.js:
```diff
@@ -142,6 +159,21 @@ export default (state = {}, action) => {
         saving: false,
         saveError: null,
       }
+    case RECEIVED_REQUEST: {
```
You can add this under `case UNPACK_REQUEST` above, then all request type actions will set the loading state. Same with the failure action below.

jspaine commented on client/modules/food/reducers/packing.js:
```diff
@@ -142,6 +159,21 @@ export default (state = {}, action) => {
+    case RECEIVED_SUCCESS:
+      return {
+        ...state,
+        ids: without(state.ids, action.response.packages._id).concat(action.response.packages),
```
The ids don't change, so all you'd need to do here is clear saving and error states

jspaine commented on client/modules/food/components/packing/Packages.js:
```diff
@@ -8,27 +8,41 @@ import {BootstrapTable, TableHeaderColumn, SizePerPageDropDown} from 'react-boot
 })
 
 class Packages extends Component {
+  constructor() {
```
If you make the changes I commented in the reducer you shouldn't need to keep any state or reload the packages

jspaine commented on server/controllers/packing.js:
```diff
@@ -14,7 +14,14 @@ export default {
   },
-
+  complete: async function(req, res) {
+    const singlePackageId = req.body.singlePackage
```
I'd just call it `packageId`, if it was more than a single one it would be an array

jspaine commented on client/modules/food/components/packing/Packages.js:
```diff
@@ -8,27 +8,41 @@ import {BootstrapTable, TableHeaderColumn, SizePerPageDropDown} from 'react-boot
 
 import selectors from '../../../../store/selectors'
-import {listPackages, unpackPackage} from '../../reducers/packing'
+import {listPackages, unpackPackage, receivedPackage} from '../../reducers/packing'
```
As the food bank, the packages are delivered not received, I'd change the names in the code and ui text

mbetances805 commented:
@jspaine I have made the changes per your feedback. Please review and let me know if any further changes are required. Thanks!

jspaine commented:
Code looks good! The buttons don't quite fit for me though and it still says 'Received' (just saw the status also says Received), want to fix it or shall I just merge this?

mbetances805 commented:
@jspaine I was following what I saw in the model for status. I can update the model to have the status as 'Delivered' instead of 'Received' and update the code elsewhere to reflect that change. I'll also look into the responsiveness of the two buttons under one column.

mbetances805 commented:
@jspaine, apologies for the delay. I updated the status and variable names from 'Received' to 'Delivered'. I addressed the responsive issue with the two buttons in one column. Please review and let me know if any further updates are required.

kenjiO commented on client/modules/food/components/packing/Packages.js:
```diff
@@ -84,9 +109,7 @@ class Packages extends Component {
-            <TableHeaderColumn
-              dataFormat={this.getActionButtons}
-            />
+            <TableHeaderColumn dataFormat={this.getActionButtons} />
```
The unpack and delivered buttons are not fitting on my browser either.  Can you change this line to `<TableHeaderColumn dataFormat={this.getActionButtons} width="200px"/>`?  That should make the column big enough so they fit side by side.

jspaine commented on client/modules/food/components/packing/Packages.js:
```diff
@@ -30,6 +32,12 @@ class Packages extends Component {
 
+  componentWillReceiveProps(nextProps) {
+    if (this.props.packSaving !== nextProps.packSaving) {
+      this.props.load()
```
You don't need to reload if you send the data back in the right format, the api middleware will handle updating the store. See below

jspaine commented on client/modules/food/reducers/packing.js:
```diff
@@ -44,6 +48,19 @@ export const pack = newPackedCustomers => dispatch => {
     .catch(error => dispatch(packFailure(error)))
 }
 
+export const deliveredPackage = singlePackage => dispatch => {
```
This can just be:
```js
export const deliveredPackage = packageId => ({
  [CALL_API]: {
    endpoint: 'packing',
    method: 'PUT',
    body: {packageId},
    responseSchema: packageSchema,
    types: [DELIVERED_REQUEST, DELIVERED_SUCCESS, DELIVERED_FAILURE]
  }
})
```

The api returns a package, it normalizes it with the package schema and updates the package entities in the store, so the ui updates.

mbetances805 replied:
@jspaine I made the updates, but it seems that I need to change the responseSchema to arrayOfPackages for the UI to update. Please confirm if you agree and I'll wrap up the updates.

jspaine replied:
Ah, it should work if you respond with single package from the controller, are you sure you changed the response in the controller to `res.json(deliveredPackage)` rather than `res.json({packages: deliveredPackage})`?

Otherwise just push it and I'll take a look. Thanks!

jspaine commented on server/controllers/packing.js:
```diff
@@ -14,7 +14,14 @@ export default {
   },
-
+  complete: async function(req, res) {
+    const packageId = req.body.singlePackage
```
`const {packageId} = req.body`

jspaine commented on server/controllers/packing.js:
```diff
@@ -14,7 +14,14 @@ export default {
+    if (!deliveredPackage) {
+      throw new BadRequestError(`package with _id ${req.body.singlePackage} not found`)
+    }
+    res.json({packages: deliveredPackage})
```
`res.json(deliveredPackage)`

mbetances805 commented:
@jspaine, made updates requested. please review.

jspaine commented:
Looks good! Thanks!
